### ISSUE TITLE
feat(cuda): cache cuDNN descriptors for repeated operations

### ISSUE THREAD

**Issue by Samyssmile** [Samyssmile/edux#219]
## Problem

`CudaOps` creates and destroys cuDNN descriptors (tensor, filter, convolution, pooling, batch norm) on **every** forward and backward call. For a typical training run with 50 epochs × 1000 batches, the same descriptors for identical shapes are created/destroyed ~50,000 times per layer.

This adds measurable CPU overhead:
- `cudnnCreateTensorDescriptor` + `cudnnSetTensor4dDescriptor` + `cudnnDestroyTensorDescriptor` per operation
- Same for filter, convolution, pooling, and activation descriptors
- Blocks the CPU thread that could instead be preparing the next kernel launch

### 1. Descriptor Cache
Implement a shape-keyed cache for cuDNN descriptors:

```java
public class CudnnDescriptorCache {
    // Key: (type, shape...) → Value: descriptor handle
    private final Map<DescriptorKey, MemorySegment> cache = new HashMap<>();
    
    public MemorySegment getTensor4d(int n, int c, int h, int w) { ... }
    public MemorySegment getFilter4d(int k, int c, int r, int s) { ... }
    public MemorySegment getConvolution(int padH, int padW, int strideH, int strideW, int dilH, int dilW) { ... }
    public MemorySegment getPooling2d(int windowH, int windowW, int padH, int padW, int strideH, int strideW, int mode) { ... }
    
    public void releaseAll() { ... } // Called on CudaOps.close()
}
```

### 2. Cache Key Design
Use a compact record as key to avoid object allocation:
```java
record DescriptorKey(int type, int d0, int d1, int d2, int d3, int d4, int d5) {}
```
This covers all descriptor types with a uniform key.

### 3. Lifecycle Management
- Cache lives in `CudaOps` instance (same lifetime as CUDA context)
- `CudaOps.close()` → `descriptorCache.releaseAll()` destroys all cached descriptors
- Bounded cache size (LRU eviction if > 256 entries) to prevent unbounded growth with dynamic shapes

### 4. Thread Safety
- Training is single-threaded per model in EDUX → simple `HashMap` sufficient
- If multi-threaded inference is added later, switch to `ConcurrentHashMap`

### 5. Scope
Descriptors to cache:
- `cudnnTensorDescriptor_t` (input, output, bias, BN mean/var)
- `cudnnFilterDescriptor_t` (conv weights)
- `cudnnConvolutionDescriptor_t` (conv config)
- `cudnnPoolingDescriptor_t` (pooling config)
- `cudnnActivationDescriptor_t` (activation config)

## Expected Impact
**5–15% overall training speedup** by reducing CPU-side overhead. More impactful for:
- Models with many layers (deep CNNs, ResNets)
- Small batch sizes (CPU overhead becomes proportionally larger)
- Combined with CUDA Graphs (future), cached descriptors are a prerequisite

## References
- [cuDNN Best Practices — Descriptor Reuse](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/)
- PyTorch: cuDNN descriptor caching in `aten/src/ATen/cudnn/Descriptors.h`
